### PR TITLE
simple tagger

### PR DESCRIPTION
**Pull request by DeNeutoy**
This is going to rely on the changes to `AllenNlpTestCase` in #14, as we now have split the `Trainer` and `Model` parameters apart, so the methods for testing models are going to change respectively.

### CODE REVIEW

File: allennlp/data/dataset_readers/sequence_tagging.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    WORD###TAG [TAB] WORD###TAG [TAB] ..... \n
+
+    and converts it into a ``Dataset`` suitable for sequence tagging.
+
+    Parameters
+    ----------
+    filename : ``str``
+    token_indexers : ``List[TokenIndexer]``, optional (default=``[SingleIdTokenIndexer()]``)
+        We use this to define the input representation for the text.  See :class:`TokenIndexer`.
+        Note that the `output` representation will always be single token IDs - if you've specified
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/sequence_tagging.py]: I think this second bit is copy pasta?

File: allennlp/data/dataset_readers/sequence_tagging.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                                           'sequence_tags': sequence_tags}))
+        return Dataset(instances)
+
+    @classmethod
+    def from_params(cls, params: Params):
+        """
+        Parameters
+        ----------
+        filename : ``str``
+        tokenizer : ``Params``, optional
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/sequence_tagging.py]: No tokenizer here.

File: allennlp/models/simple_tagger.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
+from typing import Dict
+
+import numpy
+import torch
+from torch.nn.modules.linear import Linear
+from torch.nn.modules import LSTM
+import torch.nn.functional as F
+
+from ..training.model import Model
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: Switch to absolute imports, please.

File: allennlp/models/simple_tagger.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                                                           self.num_classes))
+        self.sequence_loss = torch.nn.CrossEntropyLoss()
+
+    def forward(self,  # pylint: disable=arguments-differ
+                sequence_tokens: torch.IntTensor,
+                sequence_tags: torch.IntTensor = None):
+        """
+        Parameters
+        ----------
+        sequence_tokens : torch.LongTensor, required
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: You have `LongTensor` here, but `IntTensor` above.

File: allennlp/models/simple_tagger.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            A scalar loss to be optimised.
+
+        """
+        batch_size = sequence_tokens.size()[0]
+        # TODO(Mark): Change to use NlpApi.
+        embedded_text_input = self.embedding(sequence_tokens)
+        encoded_text, _ = self.stacked_encoders(embedded_text_input)
+
+        logits = self.tag_projection_layer(encoded_text)
+        reshaped_log_probs = logits.view(-1, self.num_classes)
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: There's no equivalent of `Dense(activation='softmax')`?  Seems like a common enough pattern that we might just want to implement a module that does this.

File: allennlp/models/simple_tagger.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        class_probabilities = F.softmax(reshaped_log_probs).view([batch_size, -1, self.num_classes])
+
+        output_dict = {"logits": logits, "class_probabilities": class_probabilities}
+
+        if sequence_tags:
+            # NLL criterion takes integer labels, not one hot.
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: "NLL criterion"?

File: allennlp/models/simple_tagger.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            if sequence_tags.dim() == 3:
+                _, sequence_tags = sequence_tags.max(-1)
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: Won't this give you a bunch of 1s?  Don't you want an `argmax`?

Is there an easy way to test the computed loss?

File: allennlp/models/simple_tagger.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        tags : List[str]
+            A list the length of the text input, containing the predicted (argmax) tag
+            from the model per token.
+        class_probabilities : numpy.Array
+            An array of shape (text_input_length, num_classes), where each row is a
+            distribution over classes for a given token in the sentence.
+        """
+        text_field.index(self.vocabulary)
+        padding_lengths = text_field.get_padding_lengths()
+        array_input = text_field.pad(padding_lengths)
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: We should probably change the `pad` methods everywhere to be `as_arrays`, as is done in the `Dataset`.  But that can be a separate PR.

File: allennlp/models/simple_tagger.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        # TODO(Mark): Generalise how the array is transformed into a variable after settling the data API.
+        # Add a batch dimension by unsqueezing, because pytorch
+        # doesn't support inputs without one.
+        array_input = torch.autograd.Variable(torch.LongTensor(array_input[0])).unsqueeze(0)
+        output_dict = self.forward(sequence_tokens=array_input)
+
+        # Remove batch dimension, as we only had one input.
+        predictions = output_dict["class_probabilities"].data.squeeze(0)
+        _, argmax = predictions.max(-1)
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: Again here, this looks like you're just doing a max, not an argmax. See http://pytorch.org/docs/master/tensors.html#torch.Tensor.max, which links to http://pytorch.org/docs/master/torch.html#torch.max.

Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: Nevermind, just read the docs more closely, and you're right.  That's confusing.  It doesn't help that the annotated return type in the docs is wrong...

File: allennlp/data/dataset_readers/sequence_tagging.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+            instances = []
+            for line in data_file:
+                tokens_and_tags = [pair.split("###") for pair in line.strip("\n").split("\t")]
+                tokens = [x[0] for x in tokens_and_tags]
+                tags = [x[1] for x in tokens_and_tags]
+
+                sequence = TextField(tokens, self._token_indexers)
+                sequence_tags = TagField(tags, sequence)
+                instances.append(Instance({'sequence_tokens': sequence,
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/sequence_tagging.py]: I'd just call these "tokens" and "tags" - having "sequence" in these names seems redundant.

File: allennlp/models/simple_tagger.py (hunk 6 above)
Comment (DeNeutoy) [inline allennlp/models/simple_tagger.py]: Losses are called criterions in torch. NLL = Negative log likelihood, i'll change this.

File: allennlp/models/simple_tagger.py (hunk 4 above)
Comment (DeNeutoy) [inline allennlp/models/simple_tagger.py]: Thanks. There are annoying differences in pytorch with functions which should take both, but don't. For instance, you can only index other tensors with `LongTensors`, not `IntTensors`.

File: allennlp/models/simple_tagger.py (hunk 5 above)
Comment (DeNeutoy) [inline allennlp/models/simple_tagger.py]: No - also we don't want to do that here, because we the cross entropy loss function takes unnormalised log probabilities for stability reasons.

Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: Fair enough.
